### PR TITLE
[DI][Bug] Autowiring thinks optional args on core classes are required

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | 3,3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | none
| License       | MIT
| Doc PR        | n/a

Currently, the following fails:

```yml
services:
    PDO:
        class: PDO
        arguments:
            - 'sqlite:/foo.db'
```

The error:

> Cannot autowire service "PDO": argument "$username" of method "__construct()" must have a type-hint or be given a value explicitly

`$username` is the second argument to `PDO`, and it's optional. Here's the reason: it appears that `$parameter->isDefaultValueAvailable()` returns false for optional arguments of core classes. But, `$parameter->isOptional()` returns true.

This allows optional arguments to not throw an exception. I can't think of any edge cases this will cause - but it's possible I'm not thinking of something :).

Cheers!

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 1]
```diff
@@ -273,6 +273,12 @@ private function autowireMethod(\ReflectionFunctionAbstract $reflectionMethod, a
 
                 // no default value? Then fail
                 if (!$parameter->isDefaultValueAvailable()) {
+                    // For core classes, isDefaultValueAvailable() can
+                    // be false when isOptional() returns true. If the
+                    // argument *is* optional, allow it to be missing
+                    if ($parameter->isOptional()) {
```
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: Why not change the `!$parameter->isDefaultValueAvailable()` condition instead?

**Review by dunglas — APPROVED** [review]

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php (hunk 1 above)
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: @fabpot it was the default behavior, but @nicolas-grekas changed it to make the pass more predictible: https://github.com/weaverryan/symfony/commit/146f07471e8876da6e92c958897a5e798954e219#diff-62df969ae028c559d33ffd256de1ac49L261

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 2]
```diff
@@ -273,6 +273,12 @@ private function autowireMethod(\ReflectionFunctionAbstract $reflectionMethod, a
...
+                        continue;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: this has to be a "break", because we cannot have holes in the parameter list

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php (hunk 1 above)
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: But yes, this change basically reverts @nicolas-grekas's change.

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: nope, it's doesn't revert it, it makes it stronger - while reading, the previous was unclear ("why does this imply that" questions) - now it's doubly clearer :)

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php (hunk 2 above)
Comment (weaverryan) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: There could be holes with `break` or `continue`. The already-specified `$arguments` are passed into this function (imagine that thanks to named arguments, someone has specified the 1st and 4th arguments to something (i.e. $arguments has`0` and `3` indexes). If we hit this code when trying to autowire the 2nd argument (index `1`), then a `break` would still mean that `$arguments` ultimately has indexes `0` and `3` (the same with a `continue`: continue would loop again, but not fill in any args).

So I think changing to `break` makes sense... but we still have holes. This is caught later by `CheckArgumentsValidityPass` with:

> Invalid constructor argument 4 for service "foo_bar": argument 2 must be defined before. Check your service definition.

tl;dr This DOES allow for holes in the parameters list in this pass... but that problem is caught by a later pass.

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: it does not *create* holes - of course, if the input has, in can only fill some of them :)

**Review by nicolas-grekas — APPROVED** [review]

**Review by xabbuh — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @weaverryan.

**Comment by nicolas-grekas** [comment]
It looks like we missed checking appveyor, where we hit a php 5.5.9 bug, see eg
https://ci.appveyor.com/project/fabpot/symfony/build/1.0.24693#L1598

which fails because $username is not optional in 5.5.9 (but it is in all following versions):
https://3v4l.org/n8ihn

**Comment by nicolas-grekas** [comment]
Fixed in #23631
